In [1]:
import glob
import os

import pandas as pd 
import numpy as np
import pprint
from sklearn.linear_model import SGDRegressor, SGDClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import log_loss, accuracy_score

In [2]:
os.environ['OCTAVE_EXECUTABLE'] = "C:/Octave/Octave-4.2.1/bin/octave-cli-4.2.1.exe"

In [3]:
%load_ext oct2py.ipython

In [4]:
from oct2py import octave
_ = octave.addpath('LOFS_Octave/source_codes/')

In [5]:
all_train = glob.glob("microarray/*_train.csv") + glob.glob("uci/*_train.csv") + glob.glob("NIPS/*_train.csv")
all_train = glob.glob("uci/*_train.csv") + glob.glob("NIPS/*_train.csv")
all_train = glob.glob("uci/*_train.csv") + glob.glob("microarray/*_train.csv")

In [6]:
def train_label(fname):
    targetname = fname.replace(".csv", ".labels")
    return pd.read_csv(targetname)

In [7]:
all_train

['uci\\Ionosphere_train.csv',
 'uci\\spambase_train.csv',
 'uci\\spectf_train.csv',
 'uci\\wdbc_train.csv',
 'microarray\\colon_train.csv',
 'microarray\\leukemia_train.csv',
 'microarray\\lung_cancer_train.csv',
 'microarray\\prostate_train.csv']

In [8]:
results_all = []
for fpath in all_train:    
    print(fpath)
    X = np.array(pd.read_csv(fpath))
    y = np.array(train_label(fpath)).flatten()
    alpha = 0.05

    X_train = np.hstack([y.reshape(-1, 1), X])
    osfs_feats = octave.fast_osfs_z(X_train, 1, alpha)
    feats_fix = [int(x-1) for x in list(np.array(osfs_feats).flatten())]
    print(len(feats_fix))
    
    # now fit and return metrics...
    X_sel = X[:, feats_fix]
    mod = SGDClassifier(loss='log', max_iter=5)
    mod.fit(X_sel, y)
    results = {
            'dataset': fpath,
            'accuracy': accuracy_score(y, mod.predict(X_sel)), 
           'logloss': log_loss(y, mod.predict_proba(X_sel)), 
           'feat_dim': len(feats_fix)}
    print(results)
    print("\n\n")
    results_all.append(results)

uci\Ionosphere_train.csv
2
{'dataset': 'uci\\Ionosphere_train.csv', 'accuracy': 0.74928774928774933, 'logloss': 1.2439850648023387, 'feat_dim': 2}



uci\spambase_train.csv
1
{'dataset': 'uci\\spambase_train.csv', 'accuracy': 0.60986742012605955, 'logloss': 0.7303239618316304, 'feat_dim': 1}



uci\spectf_train.csv
2
{'dataset': 'uci\\spectf_train.csv', 'accuracy': 0.79400749063670417, 'logloss': 0.86496149182959459, 'feat_dim': 2}



uci\wdbc_train.csv
2
{'dataset': 'uci\\wdbc_train.csv', 'accuracy': 0.62741652021089633, 'logloss': 12.868577496873577, 'feat_dim': 2}



microarray\colon_train.csv


2
{'dataset': 'microarray\\colon_train.csv', 'accuracy': 0.35483870967741937, 'logloss': 22.283081545103663, 'feat_dim': 2}



microarray\leukemia_train.csv


C:\Users\chapm\Anaconda3\envs\skrecipe\lib\site-packages\sklearn\linear_model\base.py:340: RuntimeWarning: overflow encountered in exp
  np.exp(prob, prob)


2
{'dataset': 'microarray\\leukemia_train.csv', 'accuracy': 0.63888888888888884, 'logloss': 12.472335920384413, 'feat_dim': 2}



microarray\lung_cancer_train.csv


2
{'dataset': 'microarray\\lung_cancer_train.csv', 'accuracy': 0.83425414364640882, 'logloss': 5.7246590709796719, 'feat_dim': 2}



microarray\prostate_train.csv


3
{'dataset': 'microarray\\prostate_train.csv', 'accuracy': 0.57843137254901966, 'logloss': 14.560464558638815, 'feat_dim': 3}





    [{'accuracy': 0.74928774928774933,
      'dataset': 'uci\\Ionosphere_train.csv',
      'feat_dim': 2,
      'logloss': 1.2439850648023387},
     {'accuracy': 0.60986742012605955,
      'dataset': 'uci\\spambase_train.csv',
      'feat_dim': 1,
      'logloss': 0.7303239618316304},
     {'accuracy': 0.79400749063670417,
      'dataset': 'uci\\spectf_train.csv',
      'feat_dim': 2,
      'logloss': 0.86496149182959459},
     {'accuracy': 0.62741652021089633,
      'dataset': 'uci\\wdbc_train.csv',
      'feat_dim': 2,
      'logloss': 12.868577496873577},
     {'accuracy': 0.35483870967741937,
      'dataset': 'microarray\\colon_train.csv',
      'feat_dim': 2,
      'logloss': 22.283081545103663},
     {'accuracy': 0.63888888888888884,
      'dataset': 'microarray\\leukemia_train.csv',
      'feat_dim': 2,
      'logloss': 12.472335920384413},
     {'accuracy': 0.83425414364640882,
      'dataset': 'microarray\\lung_cancer_train.csv',
      'feat_dim': 2,
      'logloss': 5.7246590709796719},
     {'accuracy': 0.57843137254901966,
      'dataset': 'microarray\\prostate_train.csv',
      'feat_dim': 3,
      'logloss': 14.560464558638815}]

In [9]:
pprint.pprint(results_all)

[{'accuracy': 0.74928774928774933,
  'dataset': 'uci\\Ionosphere_train.csv',
  'feat_dim': 2,
  'logloss': 1.2439850648023387},
 {'accuracy': 0.60986742012605955,
  'dataset': 'uci\\spambase_train.csv',
  'feat_dim': 1,
  'logloss': 0.7303239618316304},
 {'accuracy': 0.79400749063670417,
  'dataset': 'uci\\spectf_train.csv',
  'feat_dim': 2,
  'logloss': 0.86496149182959459},
 {'accuracy': 0.62741652021089633,
  'dataset': 'uci\\wdbc_train.csv',
  'feat_dim': 2,
  'logloss': 12.868577496873577},
 {'accuracy': 0.35483870967741937,
  'dataset': 'microarray\\colon_train.csv',
  'feat_dim': 2,
  'logloss': 22.283081545103663},
 {'accuracy': 0.63888888888888884,
  'dataset': 'microarray\\leukemia_train.csv',
  'feat_dim': 2,
  'logloss': 12.472335920384413},
 {'accuracy': 0.83425414364640882,
  'dataset': 'microarray\\lung_cancer_train.csv',
  'feat_dim': 2,
  'logloss': 5.7246590709796719},
 {'accuracy': 0.57843137254901966,
  'dataset': 'microarray\\prostate_train.csv',
  'feat_dim': 3,
 